### PR TITLE
[silcombine] (store (struct_gep addr) value) -> (store addr (struct v…

### PR DESCRIPTION
**Pull request by gottesmm**
…alue)) for single field structs.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by jrose-apple** [comment]
As an observer, can I ask why the new pattern is better than the old one?

**Comment by gottesmm** [comment]
@jrose-apple Semantically the two are equivalent. The difference is that at an optimizer level it is easier to promote from further up the type tree to a lower level of the type tree vs promoting from a lower level of the type tree to an upper level of the type tree. Generally the former is possible and the other is not. This PR changes the only case where it is possible to promote an available value of field in a type tree to an available value of the parent type (namely where one where all of the struct fields are single element) to one that most passes handle, the other.

**Review by shajrawi — APPROVED** [review]

**Review by aschwaighofer — APPROVED** [review]
LGTM

**Comment by jrose-apple** [comment]
Thanks for the explanation!

File: lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp [hunk 1]
```diff
@@ -585,6 +586,80 @@ SILInstruction *SILCombiner::optimizeLoadFromStringLiteral(LoadInst *LI) {
   return Builder.createIntegerLiteral(LI->getLoc(), LI->getType(), str[index]);
 }
 
+SILInstruction *SILCombiner::visitStoreInst(StoreInst *si) {
+  // These optimizations assume that lowered addresses are not enabled.
```
Comment (slavapestov) [inline lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp]: "are enabled" I think? "lowered addresses" means opaque values are not in use (or already lowered away)

**Review by slavapestov — APPROVED** [review]

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp]: I fixed it.

**Pull request merged**